### PR TITLE
Add NIRISS WFSS Notebook Examples

### PR DESCRIPTION
**Pull request by Rplesha**
This notebook checklist has been made available to us by the [the Notebooks For All team](https://github.com/Iota-School/notebooks-for-all/blob/main/resources/event-hackathon/notebook-authoring-checklist.md).
Its purpose is to serve as a guide for both the notebook author and the technical reviewer highlighting critical aspects to consider when striving to develop an accessible and effective notebook.

### The First Cell

- [x] The title of the notebook in a first-level heading (eg. `<h1>` or `# in markdown`).
- [x] A brief description of the notebook.
- [x] A table of contents in an [ordered list](https://www.markdownguide.org/basic-syntax/#ordered-lists) (`1., 2.,` etc. in Markdown).
- [x] The author(s) and affiliation(s) (if relevant).
- [x] The date first published.
- [ ] The date last edited (if relevant).
- [ ] A link to the notebook's source(s) (if relevant).

### The Rest of the Cells

- [x] There is only one H1 (`#` in Markdown) used in the notebook.
- [x] The notebook uses other [heading tags](https://www.markdownguide.org/basic-syntax/#headings) in order (meaning it does not skip numbers). 

## Text

- [x] All link text is descriptive. It tells users where they will be taken if they open the link.
- [x (hopefully)] All acronyms are defined at least the first time they are used.
- [x] Field-specific/specialized terms are used when needed, but not excessively.

## Code

- [x] Code sections are introduced and explained before they appear in the notebook. This can be fulfilled with a heading in a prior Markdown cell, a sentence preceding it, or a code comment in the code section.
- [x] Code has explanatory comments (if relevant). This is most important for long sections of code.
- [ ] If the author has control over the syntax highlighting theme in the notebook, that theme has enough color contrast to be legible.
- [x] Code and code explanations focus on one task at a time. Unless comparison is the point of the notebook, only one method for completing the task is described at a time.

### Images

- [ ] All images (jpg, png, svgs) have an [image description](https://www.w3.org/WAI/tutorials/images/decision-tree/). This could be
    - [ ] [Alt text](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/alt) (an `alt` property) 
    - [ ] [Empty alt text](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/alt#decorative_images) for decorative images/images meant to be skipped (an `alt` attribute with no value)
    - [ ] Captions
    - [ ] If no other options will work, the image is decribed in surrounding paragraphs.

- [ ] Any [text present in images](https://www.w3.org/WAI/WCAG21/Understanding/images-of-text.html) exists in a text form outside of the image (this can be alt text, captions, or surrounding text.)

### Visualizations 

**for review; I would like comments on this**
- [ ] All visualizations have an image description. Review the previous section, Images, for more information on how to add it.
- [ ] [Visualization descriptions](http://diagramcenter.org/specific-guidelines-e.html) include
    - [ ] The type of visualization (like bar chart, scatter plot, etc.)
    - [ ] Title
    - [ ] Axis labels and range
    - [ ] Key or legend
    - [ ] An explanation of the visualization's significance to the notebook (like the trend, an outlier in the data, what the author learned from it, etc.)

- [ ] All visualizations and their parts have [enough color contrast](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html) ([color contrast checker](https://webaim.org/resources/contrastchecker/)) to be legible. Remember that transparent colors have lower contrast than their opaque versions.
- [ ] All visualizations [convey information with more visual cues than color coding](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html). Use text labels, patterns, or icons alongside color to achieve this.
- [ ] All visualizations have an additional way for notebook readers to access the information. Linking to the original data, including a table of the data in the same notebook, or sonifying the plot are all options.

### CODE REVIEW

**Comment by Rplesha** [comment]
This is being reviewed by members of the NIRISS team and @camipacifici before it will be ready to be merged.

File: notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb
Comment (camipacifici) [inline notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb:106]: I think this needs a bit of explanation of how files should be stored in this directory. Like MAST downloaded them? All rate and asn in data? Rate and asn in separate directories inside data?

File: notebooks/NIRISS_WFSS_advanced/00_niriss_mast_query_data_setup.ipynb
Comment (camipacifici) [inline notebooks/NIRISS_WFSS_advanced/00_niriss_mast_query_data_setup.ipynb:320]: Will need to add this folder programmatically if it does not exist.

File: notebooks/NIRISS_WFSS_advanced/00_niriss_mast_query_data_setup.ipynb
Comment (camipacifici) [inline notebooks/NIRISS_WFSS_advanced/00_niriss_mast_query_data_setup.ipynb:362]: State the expected number of files in a comment or a markdown.

File: notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb
Comment (camipacifici) [inline notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb:72]: This cell is necessary from the outside, please make it code with a valid path to the crds folder.

File: notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb
Comment (camipacifici) [inline notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb:66]: Setting the crds path should happen before importing jwst or it will be ignored.

File: notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb
Comment (camipacifici) [inline notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb:342]: Please be a little more specific about "make some adjustments to obtain a better output source catalog".

File: notebooks/NIRISS_WFSS_advanced/00_niriss_mast_query_data_setup.ipynb
Comment (Rplesha) [reply notebooks/NIRISS_WFSS_advanced/00_niriss_mast_query_data_setup.ipynb:320]: Done! Good catch!

File: notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb
Comment (Rplesha) [reply notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb]: @camipacifici it would actually be really helpful if you could add/if you let me know I can add why you chose the parameters you did for source_catalog and tweakreg. I didn't change what you had in there before, and they look better, but I don't actually know all of the secrets. I did update the text above that call to point to some documentation about possible parameters to change (this accidentally was cut and jupyter overwrote the save before I could bring it back).

File: notebooks/NIRISS_WFSS_advanced/02_niriss_wfss_spec2.ipynb
Comment (Rplesha) [inline notebooks/NIRISS_WFSS_advanced/02_niriss_wfss_spec2.ipynb:646]: @camipacifici would you be able to provide details about the source you picked? Steph had this suggestion, and I'm not actually 100% sure what source I'm looking at or what lines there are.

**Comment by camipacifici** [comment]
I talked to the CI folks and got the confirmation that each notebook runs individually, so notebook 01 cannot use the output of notebook 00. Same deal for notebook 02. Can you please store on box the outputs of notebook 00 and 01 so that notebooks 01 and 02 can grab them? I sent you an invite to the box folder. Feel free to create a new folder for this set.

**Comment by Rplesha** [comment]
I've added the data files to the box folder `niriss_wfss_advanced`, and then they are separated in there by notebook. The notebooks expect the data in a specific directory though, not necessarily just the top level directory. Does the box folder need to match that structure to work properly?

**Comment by Rplesha** [comment]
@sosey Paul confirmed that this is true for any instrument in parallel, but it is most important for the WFSS modes. I think because of that, it would be good to have someone from NIRCam at least, and I'll leave it up to you if you want to include MIRI.(I don't think NIRSpec can be used in parallel, but I might be wrong about that).

**Comment by camipacifici** [comment]
> I've added the data files to the box folder `niriss_wfss_advanced`, and then they are separated in there by notebook. The notebooks expect the data in a specific directory though, not necessarily just the top level directory. Does the box folder need to match that structure to work properly?

Hum...I think you have to use whatever makes sense when you download in the individual notebook. Each notebook should only pull the data it needs, not everything. Maybe I am not understanding the question, sorry.

File: notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb
Comment (camipacifici) [inline notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb:577]: In my version of jdaviz (3.8.1) and Chrome browser, the markers are loaded, but they do not show on the image. I asked the devs for help.

File: notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb
Comment (camipacifici) [inline notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb:633]: Unless somebody else on the NIRISS team has suggestions, I think it is ok to just link to the docs here. I do not remember why these specific values were chosen.

File: notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb
Comment (camipacifici) [reply notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb:577]: it was a glitch...works now

File: notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb
Comment (camipacifici) [inline notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb:124]: Somewhere around here, you can download and expand the zipped file with all the rate and association files: https://data.science.stsci.edu/redirect/JWST/jwst-data_analysis_tools/niriss_wfss_advanced/niriss_wfss_advanced_01_input.zip

File: notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb
Comment (camipacifici) [inline notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb:580]: I suggest splitting this into 2 for loops, one for the images and one for the catalogs. The linking can happen between the two loops only once.

File: notebooks/NIRISS_WFSS_advanced/02_niriss_wfss_spec2.ipynb
Comment (camipacifici) [inline notebooks/NIRISS_WFSS_advanced/02_niriss_wfss_spec2.ipynb:137]: Similarly to notebook 01, you will have to download and expand the appropriate zip file:
https://data.science.stsci.edu/redirect/JWST/jwst-data_analysis_tools/niriss_wfss_advanced/niriss_wfss_advanced_02_input.zip

**Comment by Rplesha** [comment]
> > I've added the data files to the box folder `niriss_wfss_advanced`, and then they are separated in there by notebook. The notebooks expect the data in a specific directory though, not necessarily just the top level directory. Does the box folder need to match that structure to work properly?
> 
> Hum...I think you have to use whatever makes sense when you download in the individual notebook. Each notebook should only pull the data it needs, not everything. Maybe I am not understanding the question, sorry.

@camipacifici Right now all of the data are in a single directory for each notebook. However, to run the cells, the notebooks expect that data to be in certain folders, which they're not on box. Do I need to mimic the folder structure that the notebooks expect, or is it fine that the data are just in a single level folder?

File: notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb
Comment (Rplesha) [reply notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb:124]: @camipacifici is it necessary to have this in a cell that is run by default for the checks to run? I'd rather put this in a "raw" cell since it can take several minutes to download, and the 1st notebook in this series does that for them.

File: notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb
Comment (Rplesha) [reply notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb:580]: @camipacifici can you expand on why you would do this? It seems inefficient to do it that way since I have to get the catalog name from the i2d, but perhaps there's an imviz reason to do this?

Comment (Rplesha) [reply notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb:580]: I think I answered my own question trying to apply your next comment. There's a linking issue if you use sky coordinates and the markers that doesn't exist when you are using x/y pixels

File: notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb
Comment (Rplesha) [reply notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb:594]: Done. Using https://jdaviz.readthedocs.io/en/latest/imviz/import_data.html as a reference for how to do that

**Comment by mgough-970** [comment]
Sorry about the spam, I had to update the workflow a bit.  
Additionally, all notebooks require a requirements file, these notebooks will fail until one is provided in the notebooks directory.

**Comment by Rplesha** [comment]
Thanks @mgough-970 ! I was actually just working to make a requirements file now.

File: notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb
Comment (Rplesha) [reply notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb:130]: They need to be run to pass the pep8 anyway, so I just have them as normal code blocks now.

Comment (camipacifici) [reply notebooks/NIRISS_WFSS_advanced/01_niriss_wfss_image2_image3.ipynb:130]: Yea, the notebooks do not talk to each other, so you can expect that CI does NOT use the output from notebook 1 when testing notebook 2.

**Comment by camipacifici** [comment]
Thank you for all the edits, @Rplesha!!
Whenever you are ready, just put it out of draft and we'll have it go through technical review.

**Comment by Rplesha** [comment]
@camipacifici this is ready for review now whenever there is time

**Comment by camipacifici** [comment]
This is great! Thank you!

**Review by gibsongreen — APPROVED** [review]
All CI is passing, each notebook runs without issues both independently and in order. Ready to merge!

**Pull request merged**
